### ISSUE TITLE
fs.writeFile() didn't finish the write

### ISSUE THREAD

**Issue by muyinliu** [nodejs/node-v0.x-archive#25524]
node.js version: 0.12.4 (Stable)
os: Mac OS X 10.10.3
NOTE: only appear in node.js v0.12.4, the same code works fine in node.js v0.10.29(on Linux)/v0.10.38(on Mac)

I use` iconv-lite` to convert a GBK text file to UTF8 text file, here is the code:
```
var iconv = require('iconv-lite');
var fs = require('fs');

var inputPath = './input_gbk.txt';
var outputPath = './output_utf8.txt';

var buffer = fs.readFileSync(inputPath);
console.log('buffer.length: ' + buffer.length);
var decodedContent = iconv.decode(buffer, 'gbk');
console.log(decodedContent); /* decodedContent is correct and complete so I don't think the bug comes from iconv-lite */

/* write then append, write complete(686KB, use command line tool iconv also get the same output file), WHY? */
// fs.writeFileSync(inputPath, decodedContent.substring(0, 1));
// fs.appendFileSync(outputPath, decodedContent.substring(1));

/* write all with writeFileSync, but the output file is only 638KB ! WHY? */
fs.writeFileSync(outputPath, decodedContent);
```
The input file has been upload to <a href="http://pan.baidu.com/s/1boqeVnT">here</a>

**Comment by brendanashworth** [comment]
Hi @muyinliu, I have a few questions for you.

First, is this still an issue? Second, does this issue still occur when you try something new like [node v4](https://nodejs.org/en/download/)? Finally, can you reproduce the issue without using iconv-lite?

**Comment by muyinliu** [comment]
@brendanashworth 
Wait a second. I'm testing the code in Node v4.2.6.

**Comment by muyinliu** [comment]
@brendanashworth 
The code works fine in Node v4.2.6 and get the same result as Node v0.10.38 on my Mac. 
And I don't think the bug comes from iconv-lite.

**Comment by muyinliu** [comment]
@brendanashworth 
Yes, this still an issue. The code runs in Node v0.12.4 still get the incomplete write. 
It can NOT be iconv-lite's bug because `decodedContent` is the same on all these versions(v0.10.38/v0.12.4/v4.2.6). I'm sure.

**Comment by muyinliu** [comment]
@brendanashworth 
I update the file's download link. The file's SHA1 value is `cdf91466f39e6df9a96b4c1946cf6ebf4666c3f3`

**Comment by brendanashworth** [comment]
@muyinliu I can confirm the issue occurs on v0.12.4 (writing `624K` instead of `672K` from `du -sh`), looking into it now.

**Comment by muyinliu** [comment]
@brendanashworth 
[Here](http://pan.baidu.com/s/1o6YNPVW) is my incomplete output file(SHA1: `abb683fb223390afe05ff5439fff0ac19655bffb`)
Sorry, `624K` is correct. My Mac use 1000 instead of 1024 as scale value, so it says `638K`, in fact it's `638001B`

**Comment by brendanashworth** [comment]
@muyinliu what is the size of the larger output file for you / correct file (the same, right)?

**Comment by muyinliu** [comment]
@brendanashworth 
[The correct file](http://pan.baidu.com/s/1jGUSr6Q) is `685997 B`, SHA1: `9b1ae222ab22034707ee8208662cbde2d928ffe3`

**Comment by brendanashworth** [comment]
@muyinliu did you fix the typo in your test case? I'm getting SHA1: `9b1ae222ab22034707ee8208662cbde2d928ffe3` after fixing the typo.

**Comment by muyinliu** [comment]
@brendanashworth 
Sorry. The correct file's SHA1 is `9b1ae222ab22034707ee8208662cbde2d928ffe3`.

**Comment by brendanashworth** [comment]
@muyinliu the change in behavior was caused by a commit in the v0.11.7 (works in v0.11.6) release that affected the output length from `iconv-lite` — this issue is not in the `fs` module. It is fixed on the new builds (io and up), if you're able to upgrade. I don't think changes occur in v0.11 are covered for backwards compatibility, and I can't get my `git bisect` to work, so this may be where I drop you off for now.

Here is [the v0.11.7 changelog](https://github.com/nodejs/node/blob/7ea34fd62418bcfd9a454ccc43cdacd0c64a7513/CHANGELOG.md#20130821-version-0117-unstable) where the issue was introduced, and here is my test script:

``` javascript
var iconv = require('iconv-lite');
var fs = require('fs');

var decodedContent = iconv.decode(fs.readFileSync('./input_gbk.txt'), 'gbk');

var length = Buffer.byteLength(decodedContent);
console.log(process.version);
console.log('output length: ' + length + ' (correct: 685997)');
console.log(length == 685997 ? 'correct length' : 'INCORRECT length');
```

If you need a fix for this issue on the 0.12 branch, you should [open an issue on the new tracker](https://github.com/nodejs/node/issues/new). Please link to this old issue, and you're asking for a backport of a **fix** that occurred between the iojs v1.0.0 release and the node v0.11.16 release (I'm not sure what commit this is, someone will have to bisect over it). I hope I helped you out! Until then, I'm going to close _this_ issue as the repo is now an archive. Thanks for filing the bug and I'm sorry it took so long to triage the issue.
